### PR TITLE
(feat): automatically delete temporary resources

### PR DESCRIPTION
**Pull request by schristoff**
This is an attempt to solve #54 - I added OwnerReferences to some smaller resources to make sure they get cleaned up.

Unsure if we should add the `TTLSecondsAfterFinished` to that resource as it has this note:
```
/ If a Job is marked as failure, the pod has to be deleted when RestartPolicy is set to OnFailure to prevent the pod keeps restarting.
// To preserve the failed pods, the RestartPolicy needs to be set as Never. The AgentAction job will create a  new pod on retry and leave the failed ones alone.
// For more details, see the github issue: https://github.com/kubernetes/kubernetes/issues/74848#issuecomment-971487582
```

### CODE REVIEW

File: api/v1/agentconfig_types.go [hunk 1]
```diff
@@ -65,6 +65,10 @@ type AgentConfigSpec struct {
 	// +optional
 	VolumeSize string `json:"volumeSize,omitempty" mapstructure:"volumeSize,omitempty"`
 
+	// TTLSecondsAfterFinished set the time limit of the lifetime of a Job
+	// that has finished execution.
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty" mapstructure:"ttlSecondsAftterFinished,omitempty"`
```
[the remark this answers is not in the document]
Comment (troy0820) [reply api/v1/agentconfig_types.go:70]: https://github.com/getporter/operator/issues/243 Is the issue to resolve some of these things.  

In the meantime you'll have to execute the above command outside of the mage target to get the crds updated.

File: controllers/agentaction_controller.go [hunk 2]
```diff
@@ -254,6 +254,16 @@ func (r *AgentActionReconciler) createAgentVolume(ctx context.Context, log logr.
 			GenerateName: action.Name + "-",
 			Namespace:    action.Namespace,
 			Labels:       labels,
+			OwnerReferences: []metav1.OwnerReference{
```
Comment (troy0820) [inline controllers/agentaction_controller.go:257]: For later, but when you add this to the object that the controller creates, update the test to make sure that it has the owner reference on it when it gets created.

Comment (troy0820) [reply controllers/agentaction_controller.go:257]: `assert.Equal(t, pvc.OwnerReferences[0].Name, action.Name)`
when testing `TestAgentActionReconciler_createAgentVolume`

**Review by troy0820 — CHANGES_REQUESTED** [review]
So far so good.  

Just a word of caution, these types did not have owner references on them before and because the owner reference has Controller: true on them it will requeue the parent object any time this object changes.  If that behavior doesn't mess with the actual parent object we are good but I believe it will run them through the loop and cause a cascading deletion problem being that the parent object is requeued.  

**This is used for garbage collection of the controlled object and for reconciling the owner object on changes to controlled (with a Watch + EnqueueRequestForOwner).**

File: controllers/agentaction_controller.go [hunk 3]
```diff
@@ -420,8 +450,9 @@ func (r *AgentActionReconciler) createAgentJob(ctx context.Context, log logr.Log
 			},
 		},
 		Spec: batchv1.JobSpec{
-			Completions:  ptr.To(int32(1)),
-			BackoffLimit: agentCfg.GetRetryLimit(),
+			Completions:             ptr.To(int32(1)),
+			BackoffLimit:            agentCfg.GetRetryLimit(),
+			TTLSecondsAfterFinished: agentCfg.GetTTLSecondsAfterFinished(),
```
[the remark this answers is not in the document]
Comment (schristoff) [reply controllers/agentaction_controller.go:455]: If we don't call the function then we may not get the default - if there's another place to set a default (which I'm probably missing) then we can do that and nix the func

[the remark this answers is not in the document]
Comment (devigned) [reply controllers/agentaction_controller.go:455]: If you are interested in defaulting, I'd suggest looking into Kubebuilder defaults: https://book.kubebuilder.io/reference/markers.html or defaulting webhooks.

File: api/v1/agentconfig_types_test.go [hunk 4]
```diff
@@ -370,6 +370,28 @@ func TestAgentConfigSpecAdapter_GetRetryLimit(t *testing.T) {
...
 
+func TestAgentConfigSpecAdapter_GetTTLSecondsAfterFinished(t *testing.T) {
+	// var testdataDefault int32 = 600
+	var testdatauserSubmitted int32 = 700
+	testCases := []struct {
+		name                    string
+		TTLSecondsAfterFinished *int32
+		expected                *int32
+	}{
+		// {name: "default", TTLSecondsAfterFinished: nil, expected: &testdataDefault},
```
Comment (schristoff) [inline api/v1/agentconfig_types_test.go:381]: I'm going to guess that the only way to test the default that is set in the CRD is if I test it when the CRD is actually applied? :/

File: controllers/agentaction_controller.go [hunk 5]
```diff
@@ -302,6 +312,16 @@ func (r *AgentActionReconciler) createConfigSecret(ctx context.Context, log logr
 			GenerateName: action.Name + "-",
 			Namespace:    action.Namespace,
 			Labels:       labels,
+			OwnerReferences: []metav1.OwnerReference{
```
[the remark this answers is not in the document]
Comment (schristoff) [reply controllers/agentaction_controller.go:315]: There's a bunch of notes throughout the operator stating we don't use this because it has panics - I can switch it and see if the tests get upset though - maybe it's an old note?

[the remark this answers is not in the document]
Comment (troy0820) [reply controllers/agentaction_controller.go:315]: I think it is an old note.  When changing this, locally it didn't do anything bad.

File: api/v1/agentconfig_types_test.go [hunk 6]
```diff
@@ -370,6 +370,28 @@ func TestAgentConfigSpecAdapter_GetRetryLimit(t *testing.T) {
...
+		{name: "user submitted", TTLSecondsAfterFinished: &testdatauserSubmitted, expected: &testdatauserSubmitted},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			adapter := NewAgentConfigSpecAdapter(AgentConfigSpec{
+				TTLSecondsAfterFinished: tc.TTLSecondsAfterFinished,
+			})
+			result := adapter.GetTTLSecondsAfterFinished()
```
Comment (troy0820) [inline api/v1/agentconfig_types_test.go:389]: This is testing that your getter and your setter work.  

I'm not sure but if you test the creation of this resource without setting it does it show that this is set with the default? _(Probably not unit wise but with maybe env test, because env test uses a fake api-server and loads the crds)_

What worries me is this:
You have an `agentConfig`, you have an `agentConfigAdapter`, you have an `agentConfigSpecAdapter`. 

If you are putting this on the agentConfig, those things will get set by the nature of them copying the existing structure over to the new "type".  Satisfying the other types, _(yes you need the Get method to get the TTL set on the original agentconfig)_ makes this more complicated that it has to be.  I would imagine that the api/v1 type (the agentconfig) is the one with the owner reference and needs to have that information on it to pass to the batch/v1 job to settle the timeout. The adapter extending this doesn't seem consistent with what  putting the type on the agentConfig. 

TLDR; to not change anything with the crds, you can just use the non exposed api/v1 surface _(agenconfigSpecAdapter)_ and you can perform this same functionality without changing the schema for the api that exists.  That way you don't have to do anything with agentconfig besides add the owner on it.

Comment (troy0820) [reply api/v1/agentconfig_types_test.go:389]: So it may be better served to not add it to the agentconfig type but to the agentconfigspecAdapter type.

File: api/v1/zz_generated.deepcopy.go [hunk 7]
```diff
@@ -249,6 +249,11 @@ func (in *AgentConfigList) DeepCopyObject() runtime.Object {
 // DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
 func (in *AgentConfigSpec) DeepCopyInto(out *AgentConfigSpec) {
 	*out = *in
+	if in.TTLSecondsAfterFinished != nil {
```
Comment (troy0820) [inline api/v1/zz_generated.deepcopy.go:252]: With the above suggestion, you will not have to do this.

**Review by troy0820 — CHANGES_REQUESTED** [review]
Just a couple of changes, and the testing effort may need to be done in an env test or some integration that can show that the resources are cleaned up and the default value is set.

File: controllers/agentaction_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (schristoff) [reply controllers/agentaction_controller.go:315]: kk will update

**Comment by schristoff** [comment]
Closing due to breaking this apart
